nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/PutCassandraRecord.java

```diff
@@ -0,0 +1,222 @@
+        if (inputFlowFile == null) {
+            return;
+        }
+
+        final String cassandraTable = context.getProperty(TABLE).evaluateAttributeExpressions(inputFlowFile).getValue();
+        final RecordReaderFactory recordParserFactory = context.getProperty(RECORD_READER_FACTORY).asControllerService(RecordReaderFactory.class);
+        final int batchSize = context.getProperty(BATCH_SIZE).evaluateAttributeExpressions().asInteger();
+        final String batchStatementType = context.getProperty(BATCH_STATEMENT_TYPE).getValue();
+        final String serialConsistencyLevel = context.getProperty(CONSISTENCY_LEVEL).getValue();
+
+        final BatchStatement batchStatement;
+        final Session connectionSession = cassandraSession.get();
+        final AtomicInteger recordsAdded = new AtomicInteger(0);
+        final StopWatch stopWatch = new StopWatch(true);
+
+        boolean error = false;
+
+        try (final InputStream inputStream = session.read(inputFlowFile);
+             final RecordReader reader = recordParserFactory.createRecordReader(inputFlowFile, inputStream, getLogger())){
+
+            final RecordSchema schema = reader.getSchema();
+            Record record;
+
+            batchStatement = new BatchStatement(BatchStatement.Type.valueOf(batchStatementType));
+            batchStatement.setSerialConsistencyLevel(ConsistencyLevel.valueOf(serialConsistencyLevel));
+
+            while((record = reader.nextRecord()) != null) {
+                Map<String, Object> recordContentMap = (Map<String, Object>) DataTypeUtils
+                        .convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema()));
+                Insert insertQuery;
+
+                if (cassandraTable.contains(".")) {
+                    String keyspaceAndTable[] = cassandraTable.split("\\.");
+                    insertQuery = QueryBuilder.insertInto(keyspaceAndTable[0], keyspaceAndTable[1]);
+                } else {
+                    insertQuery = QueryBuilder.insertInto(cassandraTable);
+                }
+                for (String fieldName : schema.getFieldNames()) {
+                    insertQuery.value(fieldName, recordContentMap.get(fieldName));
+                }
+                batchStatement.add(insertQuery);
+
+                if (recordsAdded.incrementAndGet() == batchSize) {
+                    connectionSession.execute(batchStatement);
+                    batchStatement.clear();
+                }
+            }
+
+            if (batchStatement.size() != 0) {
+                connectionSession.execute(batchStatement);
+                batchStatement.clear();
+            }
+
+        } catch (Exception e) {
+            error = true;
+            getLogger().error("Unable to write the records into Cassandra table due to {}", new Object[] {e});
+            session.transfer(inputFlowFile, REL_FAILURE);
+        } finally {
+            if (!error) {
+                stopWatch.stop();
```

Comment:
Should have been more specific. What I was saying is why not put all of this transfer logic inside of the `try` segment. As in right after the block that starts on line 190 (`if (batchStatement.size()...`)

Reply:
So it would follow this pattern:

```
try {
    // Do stuff with the batch writing...
    // Now send input flowFile to REL_SUCCESS
} catch (Exception ex) {
    //LOG error
    //send input flowFile to REL_FAILURE
}
```

Cuz I just don't think we need a `finally` block here.